### PR TITLE
add option to dock terminal at top

### PR DESCRIPTION
**Pull request by mxts**
for https://github.com/microsoft/vscode/issues/201452

### CODE REVIEW

**Comment by mxts** [comment]
@abhijit-chikane would you be so kind to help with community PR approvals? thanks!

**Review by abhijit-chikane — APPROVED** [review]

**Comment by bilogic** [comment]
Thanks! Now we just need 1 more approval.

**Comment by mxts** [comment]
@sbatten would you be able to help? Thank you.

**Review by bilogic — APPROVED** [review]

**Comment by mxts** [comment]
@sbatten can help? thanks

**Comment by mxts** [comment]
@benibenj @sbatten @meganrogge possible to give some feedback on this please? 

I would have opt for an extension if it was possible to move the terminal but VSCode has a restriction here https://code.visualstudio.com/api/extension-capabilities/overview#no-dom-access

I use a 27" monitor, without this feature, I constantly have to look up and down from the top section to the lowest line on the screen and really causes strain on the neck. Thanks.

**Comment by benibenj** [comment]
Have you already tried moving the terminal into the editor? This would allow you to place the terminal at the top.
image

**Comment by mxts** [comment]
@benibenj how do you do that? Would it activate with `ctrl`+`~`?

**Comment by mxts** [comment]
@ agree

**Comment by mxts** [comment]
@ agree company="Avexsoft"

**Comment by benibenj** [comment]
Right click on one of the terminal instances. 

image

If you only have one terminal instance you can just left click on it in the panel bar.

image

**Comment by mxts** [comment]
Thanks, managed to move it, but `ctrl`+`~` does not activate it, it opens a new one at the bottom😅

I mean, it isn't just the terminal, it's everything in the panel, i.e. problems, output etc, terminal is just the most frequently used.

**Review by bilogic — APPROVED** [review]

**Comment by mxts** [comment]
@benibenj i gave your suggestions a shot, but it created other problems, is a merge possible? Thank you.

**Comment by benibenj** [comment]
I'll have a look at this in June and discuss with the team if this is something we want to support

**Comment by bilogic** [comment]
thanks @benibenj, fyi, there is actually > 20 votes here https://github.com/microsoft/vscode/issues/50984

**Comment by heartacker** [comment]
any updates? I really appreciate the idea

**Comment by benibenj** [comment]
Moving this to July. We want to support having the terminal and other views at the top, we just need to decide if the approach used in this PR is what we want to go for.

**Comment by benibenj** [comment]
@mxts Can you resolve the conflict and pull from main. Would like to get this PR merged this week.

**Comment by mxts** [comment]
@benibenj done!

**Comment by benibenj** [comment]
There is still a couple things missing:
- Panel Border bottom when it is at the top:
image
=> in `panelpart.css`: 
```
.monaco-workbench .part.panel.top {
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

.monaco-workbench.nomaineditorarea .part.panel.top {
	border-bottom-width: 0; /* no border when main editor area is hiden */
}
```
=> in `panelPart.ts`: 
```
updateStyles(){
        ...
        container.style.borderBottomColor = borderColor;
        ...
}

override layout(width: number, height: number, top: number, left: number): void {
		let dimensions: Dimension;
		switch (this.layoutService.getPanelPosition()) {
			case Position.RIGHT:
				dimensions = new Dimension(width - 1, height); // Take into account the 1px border when layouting
				break;
			case Position.TOP:
				dimensions = new Dimension(width, height - 1); // Take into account the 1px border when layouting
				break;
			default:
				dimensions = new Dimension(width, height);
				break;
		}

		// Layout contents
		super.layout(dimensions.width, dimensions.height, top, left);
	}
```

- Show correct maximize icon:
image

- Support for aligning the panel to left/right when at the top
Code_-_OSS_A8vOAfz5ih

- When reloading the panel is rendered at the bottom

File: src/vs/workbench/browser/parts/editor/editorPart.ts
Comment (benibenj) [inline src/vs/workbench/browser/parts/editor/editorPart.ts:1127]: `!` is not needed

File: src/vs/workbench/browser/parts/editor/editorPart.ts
Comment (benibenj) [inline src/vs/workbench/browser/parts/editor/editorPart.ts:1122]: the `!` you added is not needed

**Review by benibenj — CHANGES_REQUESTED** [review]

**Comment by benibenj** [comment]
@mxts would you be open to grant me permissions to push commits to this PR? This would allow me to help you out with some of the required changes. There should be a button called `Allow edits from maintainers`

**Comment by mxts** [comment]
Certainly! But how do I do that? I'm trying to look for allow maintainers to edit

**Comment by benibenj** [comment]
You should be able to see some action/checkbox below `reviewers` on this page 
Docs: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by mxts** [comment]
image

I just don't have it, an alternative is that i invited to avexsoft.. maybe then u can commit to it.

**Comment by benibenj** [comment]
I guess there is not mention of it further down in the sidebar? I just found online that it would be bellow the list of participants

image

image

**Comment by mxts** [comment]
image

Nothing

**Comment by mxts** [comment]
Ok, let me know if there is any settings I can change.

**Comment by benibenj** [comment]
Let's try this: I created a PR with my changes against your fork. If you review and approve it, it will be part of this PR.

https://github.com/avexsoft/vscode/pull/1

**Comment by mxts** [comment]
I have merged the changes, but have some trouble compiling VSCode... let me try and sort it out.

**Review by benibenj — APPROVED** [review]

**Comment by benibenj** [comment]
I approved it. This will allow the pipeline to run. Let me know after you tested and are happy with the behaviour. If all tests pass and you are happy with my changes, we can merge the PR :)

**Review by bpasero — APPROVED** [review]

**Review by Tyriar — APPROVED** [review]
terminal/ looks good 👏

**Comment by mxts** [comment]
yes @benibenj please merge thanks!

**Pull request merged**
